slick/src/main/scala/slick/util/AsyncExecutor.scala

```diff
@@ -62,141 +76,201 @@ object AsyncExecutor extends Logging {
+
+      case n =>
+        // NOTE: The current implementation of ManagedArrayBlockingQueue is flawed. It makes the assumption that all
+        // tasks go through the queue (which is responsible for scheduling high-priority tasks first). However, that
+        // assumption is wrong since the ThreadPoolExecutor bypasses the queue when it creates new threads. This
+        // happens whenever it creates a new thread to run a task, i.e. when minThreads < maxThreads and the number
+        // of existing threads is < maxThreads.
+        //
+        // The only way to prevent problems is to have minThreads == maxThreads when using the
+        // ManagedArrayBlockingQueue.
+        require(minThreads == maxThreads, "When using queueSize > 0, minThreads == maxThreads is required.")
+
+        // NOTE: The current implementation of ManagedArrayBlockingQueue.increaseInUseCount implicitly `require`s that
+        // maxThreads <= maxConnections.
+        require(maxThreads <= maxConnections, "When using queueSize > 0, maxThreads <= maxConnections is required.")
+
+        // NOTE: Adding up the above rules
+        // - maxThreads >= maxConnections, to prevent database locking issues when using transactions
+        // - maxThreads <= maxConnections, required by ManagedArrayBlockingQueue
+        // - maxThreads == minThreads, ManagedArrayBlockingQueue
+        //
+        // We have maxThreads == minThreads == maxConnections as the only working configuration
+
+        val managedArrayBlockingQueue = new ManagedArrayBlockingQueue(maxConnections, n)
+
+        new AsyncExecutorImpl(managedArrayBlockingQueue.asInstanceOf[BlockingQueue[Runnable]]) {
+          override def prioritizedRunnable(priority: => Priority,
+                                           run: PrioritizedRunnable.SetConnectionReleased => Unit
+                                          ): PrioritizedRunnable = {
+            def _priority = priority
+            def _run(setConnectionReleased: PrioritizedRunnable.SetConnectionReleased) =
+              run(setConnectionReleased)
+            new PrioritizedRunnable {
+              override def priority() = _priority
+              override def run() =
+                if (priority() != WithConnection)
+                  if (managedArrayBlockingQueue.attemptPrepare(this)) {
+                    try _run(new PrioritizedRunnable.SetConnectionReleased(() => connectionReleased = true))
+                    finally
+                    // If the runnable/task has released the Jdbc connection we decrease the counter again
+                      if (!managedArrayBlockingQueue.attemptCleanUp(this))
+                        logger.warn("After executing a task, the in-use count was already 0. This should not happen.")
+                  } else {
+                    logger.warn("Could not increase in-use count. Will resubmit...")
+                    executor.execute(this)
+                  }
```

Comment:
We also need to `_run` it when `priority() == WithConnection`

Reply:
Silly me. Good catch, fixed